Add Documentation after 0.3.0 release (#94)
AbsaOSS/hermes: _docs/0.3/usage/e2e-runner.md

---
layout: docs
title: Usage - E2E Runner
version: '0.3'
categories:
    - '0.3'
    - usage
redirect_from: /docs/usage/dataset-comparison
---

#### Table Of Contents
- [Table Of Contents](#table-of-contents)
- [Properties](#properties)
- [Parameters](#parameters)
- [Test Definition JSON](#test-definition-json)

#### Properties

> All properties were scraped. Only properties that might come in hadny are the ones implemented atop scribe logging library. Check them out on their GitHub [page](https://github.com/outr/scribe)

Reviewer: The meaning does not seem to be clear to me. Do you mean _scrapped_ as in _shredded_? But if so, this does not seem proper description terms, why not simply "There are no properties to be defined"?

Or do you actually mean _scraped_ from somewhere? (I am not familiar with Scribe) If so, maybe this would deserve some more description on how this is can be used?

Reviewer 2: I am little confused here as well. I though the implicit reference (of scrapped properties) was to previous version of Hermes. 🤔

Author: I just left a comment about scribe. No need for more

> All properties were scraped. Only properties that might come in hadny are the ones implemented atop scribe logging library. Check them out on their GitHub [page](https://github.com/outr/scribe)

Reviewer 2: hadny -> handy

Author: Fixed

#### Parameters

> - `test-definition-path` - mandatory - path to json with test definitions. Format explanation bellow in [Test Definition JSON](#test-definition-json)

Reviewer 2: Really minor:
```suggestion
- `test-definition-path` - mandatory - path to json with test definitions. Format explanation bellow in [Test Definition JSON](#test-definition-json).
```

- `jar-path` - optional - path to additional jars to load plugins from. You can also use spark's `--jars` instead of this.
- `fail-fast` - defaults to `false` - switch in case you want to kill the run on first failure.

#### Test Definition JSON

Test definition JSON has a struct as base having two keys, `vars` and `runs`. 

> `vars` holds a struct used for representation of variables inside of runs. This is so you do not have to repeat yourself with, for example, dataset names or other parameters that are valid for more then one run. Key is the variable you put in other places of the JSON surounded by `#{` and `}#` and the value is the value you want to use in that place.

Reviewer: ```suggestion
`vars` holds a struct used for the representation of variables for multiple runs. This way, you do not have to repeat yourself with, for example, dataset names or other parameters that are valid for more than one run. Key is the variable you put in other places of the JSON surrounded by `#{` and `}#` and the value is the value you want to use in that place.
```

> `runs` holds an array of test definitions. Test definition is an object, having it's own case class TestDefinition holding the information about of test run. That is a plugin name, input arguments, order, etc. 

Reviewer: ```suggestion
`runs` holds an array of test definitions. Test definition is an object, having its own `case class TestDefinition` holding the information about a test run, i.e. a plugin name, input arguments, order, etc. 
```

**JSON Structure**

```json
{
  // Struct holding the variables. Optional
  "vars": {
    // Here `prefix` is a key that will be somewhere in runs as `#{prefix}#` and the 

>     // value will be inputed there instead of it

Reviewer 2: ```suggestion
    // value will be input there instead of it
```

Author: I am not sure I agree. But I see that I m missing a `t`

Reviewer 2: OK, didn't know that _inputted_ is a valid past tense to _input_ too. Always thought it was only _input_. According to Internets both is valid. 😉

Author: Fixed

>     "prefix": "Some Random PreFIX stuff"

Reviewer 2: Sorry I don't get the description for either `vars` and `prefix` her. Particularly the later.

Author: Could we have a sit down to discuss a better description?

  },
  // Struct holding array of test definitons. Mandatory
  "runs" : [
    // Representation of one TestDefinition
    {
      // PluginName is a unique name of plugin loaded on start. Each plugin has its own specific name
      "pluginName" : "BashPlugin",
      // Name of the test. Also should be unique. Used by testers to differentiate runs
      "name": "Test2",
      // Order of the run. Order is used to sort runs by. If there are multiple same order elements,
      // name is used to sort them further in alphabetical order.
      "order" : 1,
      // Input arguments for the plugin. First argument will be replaced by var `prefix`.
      // Second is a simple string
      "args" : ["#{prefix}#", "fileName"],
      // Arguments, but for execution of write of plugin result. Optional
      "writeArgs": [],
      // If the test depends on a different test. Will automatically fail if dependee failed. Optional
      "dependsOn": "Test1"
    }
  ]
}

> ```

Reviewer 2: I think a concrete example would be most helpful here.

Reviewer 3: Totally agree. This is too general to understand this kind of plugin.
